posix: rwlock: use option group for kconfig option name

RomainPelletant opened the pull request:
- posix: options: replace kconfig for rw lock
- doc: posix: update doc for POSIX_RW_LOCKS
- doc: release: 4.3: deprecate CONFIG_POSIX_READER_WRITER_LOCKS

Fixes #81326

Commit:
posix: options: replace kconfig for rw lock

- Update Kconfig from CONFIG_POSIX_READER_WRITER_LOCKS to POSIX_RW_LOCKS

Signed-off-by: Romain Pelletant <romainp@kickmaker.net>

Commit:
doc: posix: update doc for POSIX_RW_LOCKS

- Replace POSIX_RW_LOCKS in documentation

Signed-off-by: Romain Pelletant <romainp@kickmaker.net>

cfriedt reviewed (changes requested):
Thanks - this is partially correct.

RW locks are one of those wonderful areas that have an Option Group as well as an Option, and they don't match.

Option Group (`POSIX_RW_LOCKS`)
https://pubs.opengroup.org/onlinepubs/9699919799/xrat/V4_subprofiles.html

Option (`_POSIX_READER_WRITER_LOCKS`)
https://pubs.opengroup.org/onlinepubs/9699919799/basedefs/unistd.h.html

So we do want to update the Kconfig option to match the Option Group name (`POSIX_RW_LOCKS` -> `CONFIG_POSIX_RW_LOCKS`) but the Option was correct in a few places.

We very likely want to update the deprecation as well, which will be fun 🫠😅, and yes, this is a bug.

I could go over all changes again if you like and point things out, but it might just make more sense to resubmit before I review.

I wish I would have noticed this sooner - thanks for your help!

RomainPelletant commented:
Oh I didn't get the point 😅 I will resubmit it in the right direction

cfriedt commented:
@RomainPelletant - are you able to rework your change and resubmit?

RomainPelletant commented:
@cfriedt yes, I will do it beginning of January if the deadline is OK for you

RomainPelletant commented:
@cfriedt just to be sure: could you confirm that the first modification shall be changing POSIX_READER_WRITER_LOCKS to POSIX_RW_LOCKS according to https://pubs.opengroup.org/onlinepubs/9699919799/xrat/V4_subprofiles.html?
From that confirmation I think the remaining tasks will be smooth :)

cfriedt commented:
@RomainPelletant - that sounds about right.

cfriedt commented on doc/services/portability/posix/conformance/index.rst:
```diff
@@ -40,7 +40,7 @@ POSIX System Interfaces
     :ref:`_POSIX_MAPPED_FILES<posix_option_group_mapped_files>`, 200809L, :kconfig:option:`CONFIG_POSIX_MAPPED_FILES`
     :ref:`_POSIX_MEMORY_PROTECTION<posix_option_group_memory_protection>`, 200809L, :kconfig:option:`CONFIG_POSIX_MEMORY_PROTECTION` :ref:`†<posix_undefined_behaviour>`
-    :ref:`_POSIX_READER_WRITER_LOCKS<posix_option_reader_writer_locks>`, 200809L, :kconfig:option:`CONFIG_POSIX_READER_WRITER_LOCKS`
+    :ref:`_POSIX_RW_LOCKS<posix_option_reader_writer_locks>`, 200809L, :kconfig:option:`CONFIG_POSIX_RW_LOCKS`
```
Again, I'm sorry the spec has this inconsistency but the Option has a slightly different notation than the Option Group, and the Kconfig value should copy the Option Group name.

```suggestion
    :ref:`_POSIX_READER_WRITER_LOCKS<posix_option_reader_writer_locks>`, 200809L, :kconfig:option_group:`CONFIG_POSIX_RW_LOCKS`
```

cfriedt commented on doc/services/portability/posix/option_groups/index.rst:
```diff
@@ -845,12 +845,14 @@ Enable this option with :kconfig:option:`CONFIG_POSIX_RAW_SOCKETS`.
 .. _posix_option_reader_writer_locks:
 
-_POSIX_READER_WRITER_LOCKS
+_POSIX_RW_LOCKS
 ++++++++++++++++++++++++++
 
-Enable this option with :kconfig:option:`CONFIG_POSIX_READER_WRITER_LOCKS`.
+Enable this option with :kconfig:option:`CONFIG_POSIX_RW_LOCKS`.
 
-.. csv-table:: _POSIX_READER_WRITER_LOCKS
+.. note:: POSIX_READER_WRITER_LOCKS and PTHREAD_RWLOCK are deprecated.
+
+.. csv-table:: _POSIX_RW_LOCKS
```
This whole block should be moved to where the other Option Groups are located, right after [POSIX_PIPE](https://docs.zephyrproject.org/latest/services/portability/posix/option_groups/index.html#posix-pipe) to keep things in lexographic order.

Additionally, the POSIX Option Group is `POSIX_RW_LOCKS` (not leading underscore) and the POSIX Option is `_POSIX_READER_WRITER_LOCKS` (with leading underscore).

In this block, we want to change it to the Option Group rather than use the Option notation, and that would need to include the .rst anchor.

cfriedt commented on include/zephyr/posix/posix_features.h:
```diff
@@ -99,8 +99,8 @@
-#ifdef CONFIG_POSIX_READER_WRITER_LOCKS
-#define _POSIX_READER_WRITER_LOCKS _POSIX_VERSION
+#ifdef CONFIG_POSIX_RW_LOCKS
+#define _POSIX_RW_LOCKS _POSIX_VERSION
```
```suggestion
#define _POSIX_READER_WRITER_LOCKS _POSIX_VERSION
```

cfriedt commented on include/zephyr/posix/sys/sysconf.h:
```diff
@@ -172,7 +172,7 @@ enum {
 #define __z_posix_sysconf_SC_RE_DUP_MAX _POSIX_RE_DUP_MAX
 #define __z_posix_sysconf_SC_READER_WRITER_LOCKS                                                   \
-	COND_CODE_1(CONFIG_POSIX_READER_WRITER_LOCKS, (_POSIX_READER_WRITER_LOCKS), (-1L))
+	COND_CODE_1(CONFIG_POSIX_RW_LOCKS, (_POSIX_RW_LOCKS), (-1L))
```
```suggestion
	COND_CODE_1(CONFIG_POSIX_RW_LOCKS, (_POSIX_READER_WRITER_LOCKS), (-1L))
```

cfriedt reviewed (dismissed):
Just a few slight changes left.

This is one of those annoying Option Groups with a mismatched Option name.

cfriedt commented:
@RomainPelletant - can you rebase on `main` and resolve conflicts? I think this should probably go in with the few changes suggested.

RomainPelletant commented:
@cfriedt should I create a new role for kconfig:option_group?

cfriedt commented:
@RomainPelletant - no need to create a new role in restructured text.

Commit:
doc: release: 4.3: deprecate CONFIG_POSIX_READER_WRITER_LOCKS

The Kconfig option `CONFIG_POSIX_READER_WRITER_LOCKS` is being deprecated
and renamed to `CONFIG_POSIX_RW_LOCKS` in order to match the name of the
POSIX Option Group.

`_POSIX_READER_WRITER_LOCKS` is the name of the feature test macro.

There are a few instances where the name of the Option Group does not
correspond exactly to the name of the Option (Feature Test Macro), and
this is one of them.

Signed-off-by: Chris Friedt <cfriedt@tenstorrent.com>
